packages/navie/src/commands/observe-command.ts

```diff
@@ -0,0 +1,143 @@
+import { z } from 'zod';
+
+import type Command from '../command';
+import type { CommandRequest } from '../command';
+import CompletionService, { ModelType } from '../services/completion-service';
+import LookupContextService from '../services/lookup-context-service';
+import VectorTermsService from '../services/vector-terms-service';
+import { ContextV2 } from '../context';
+import { ExplainOptions } from './explain-command';
+import Message from '../message';
+
+const RelevantTest = z.object({
+  relevantTest: z
+    .object({
+      name: z.string().optional().describe('The name of the test case, if known'),
+      path: z.string().describe('The path to the test file'),
+      language: z
+        .enum(['ruby', 'python', 'java', 'javascript', 'other'])
+        .describe('The programming language of the test file'),
+      framework: z.string().optional().describe('The test framework used'),
+    })
+    .optional()
+    .describe('A descriptor of the most relevant test to the requested behavior'),
+});
+
+export default class ObserveCommand implements Command {
+  constructor(
+    private readonly options: ExplainOptions,
+    private readonly completionService: CompletionService,
+    private readonly lookupContextService: LookupContextService,
+    private readonly vectorTermsService: VectorTermsService
+  ) {}
+
+  getTestSnippets(vectorTerms: string[], tokenLimit: number): Promise<ContextV2.FileContextItem[]> {
+    return this.lookupContextService
+      .lookupContext(vectorTerms, tokenLimit, {
+        include: ['test', 'spec'],
+      })
+      .then((contextItems) =>
+        contextItems.filter(
+          (item): item is ContextV2.FileContextItem =>
+            item.type === ContextV2.ContextItemType.CodeSnippet
+        )
+      );
+  }
+
+  async getMostRelevantTest(
+    userRequest: string,
+    fileContextItems: ContextV2.FileContextItem[]
+  ): Promise<z.infer<typeof RelevantTest>['relevantTest']> {
+    const context = fileContextItems.map(
+      ({ content, location }) =>
+        `<code-snippet location="${location}">\n${content}\n</code-snippet>`
+    );
+    const prompt = `Given the following code snippets, identify the single most relevant test to the user request.
+<code-snippets>
+${context.join('\n')}
+</code-snippets>`;
+
+    const response = await this.completionService.json(
+      [
+        {
+          role: 'system',
+          content: prompt,
+        },
+        {
+          role: 'user',
+          content: userRequest,
+        },
+      ],
+      RelevantTest,
+      { model: ModelType.Mini }
+    );
+
+    return response?.relevantTest;
```

Comment:
Is this expected to be a test file or a specific test?

There can be a lot of different syntax for how to run a specific test. So maybe we just go with the whole file?

Reply:
It returns a path to the file, so the command will typically run the entire file. There's an additional field for the name of the test, though it's optional, so sometimes the LLM will opt to suggest a command with a test filter to that name.